presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionarySelectiveReader.java

```diff
@@ -0,0 +1,602 @@
+                    int currentPosLength = dictionaryBlock.getSliceLength(index);
+                    Slice data = dictionaryBlock.getSlice(index, 0, currentPosLength);
+                    if (isCharType() && length != currentPosLength) {
+                        data = Chars.padSpaces(data, streamDescriptor.getOrcType().getLength().get());
+                    }
+                    if (filter.testBytes(data.getBytes(), 0, length)) {
+                        if (outputRequired) {
+                            values[outputPositionCount] = index;
+                        }
+                        outputPositions[outputPositionCount] = position;
+                        outputPositionCount++;
+                    }
+                }
+            }
+
+            streamPosition++;
+
+            if (filter != null) {
+                outputPositionCount -= filter.getPrecedingPositionsToFail();
+                int succeedingPositionsToFail = filter.getSucceedingPositionsToFail();
+                if (succeedingPositionsToFail > 0) {
+                    int positionsToSkip = 0;
+                    for (int j = 0; j < succeedingPositionsToFail; j++) {
+                        i++;
+                        int nextPosition = positions[i];
+                        positionsToSkip += 1 + nextPosition - streamPosition;
+                        streamPosition = nextPosition + 1;
+                    }
+                    skip(positionsToSkip);
+                }
+            }
+        }
+        return streamPosition;
+    }
+
+    private boolean isCharType()
+    {
+        return streamDescriptor.getOrcType().getOrcTypeKind() == CHAR;
+    }
+
+    private int readAllNulls(int[] positions, int positionCount)
+            throws IOException
+    {
+        presentStream.skip(positions[positionCount - 1]);
+
+        if (nonDeterministicFilter) {
+            outputPositionCount = 0;
+            for (int i = 0; i < positionCount; i++) {
+                if (filter.testNull()) {
+                    outputPositionCount++;
+                }
+                else {
+                    outputPositionCount -= filter.getPrecedingPositionsToFail();
+                    i += filter.getSucceedingPositionsToFail();
+                }
+            }
+        }
+        else if (nullsAllowed) {
+            outputPositionCount = positionCount;
+        }
+        else {
+            outputPositionCount = 0;
+        }
+
+        allNulls = true;
+        return positions[positionCount - 1] + 1;
+    }
+
+    private void ensureValuesCapacity(int capacity)
+    {
+        values = ensureCapacity(values, capacity);
+    }
+
+    private void skip(int items)
+            throws IOException
+    {
+        if (presentStream != null) {
+            int dataToSkip = presentStream.countBitsSet(items);
+            if (inDictionaryStream != null) {
+                inDictionaryStream.skip(dataToSkip);
+            }
+            if (dataStream != null) {
```

Comment:
dataStream here can't be null, can it?

Reply:
if all the elements in a array are of length 0 then the datastream is null and lengthstream is present

Reply:
@bhhari This is a dictionary reader, so values are integers and therefore there are no lengths and a value cannot be of length 0. Am I missing something?

Reply:
> @bhhari This is a dictionary reader, so values are integers and therefore there are no lengths and a value cannot be of length 0. Am I missing something?

that's right. mixed up direct. thanks